app/controllers/users_controller.rb

```diff
@@ -189,6 +189,28 @@ def disable_totp
     redirect_back_or_to security_user_path(@user)
   end
 
+  def generate_backup_codes
+    @user = params[:id] ? User.friendly.find(params[:id]) : current_user
+    authorize @user
+    @previewed_backup_codes = @user.generate_backup_codes!
+  end
+
+  def activate_backup_codes
+    @user = params[:id] ? User.friendly.find(params[:id]) : current_user
+    authorize @user
+    @user.active_backup_codes&.map(&:mark_discarded!)
+    @user.backup_codes&.previewed&.map(&:mark_active!)
```

Comment:
1. **concern**: We don't need as many `&` here. These methods return instances of `ActiveRecord::Relation` which always behaves like an array (meaning `.map` is a no-op)

    ```suggestion
        @user.active_backup_codes.map(&:mark_discarded!)
        @user.backup_codes.previewed.map(&:mark_active!)
    ```
2. **concern**: This feels like the kind of logic that
    1. Should be wrapped in a transaction so the entire operation either commits or fails
    2. Does not belong in the controller (maybe `@user.activate_new_backup_codes!` or something to that effect)

Reply:
done!